File: tests/auth.test.ts
import { describe, expect, test, afterEach } from "bun:test";
import { resolveCredentials } from "../src/auth.ts";

describe("resolveCredentials", () => {
  const originalEnv = { ...process.env };

  afterEach(() => {
    process.env.JENKINS_USER_ID = originalEnv.JENKINS_USER_ID;
    process.env.JENKINS_API_TOKEN = originalEnv.JENKINS_API_TOKEN;
  });

  test("auth flag takes priority", async () => {
    process.env.JENKINS_USER_ID = "envuser";
    process.env.JENKINS_API_TOKEN = "envtoken";
    const creds = await resolveCredentials("https://ci.example.com", "flaguser:flagtoken");
    expect(creds).toEqual({ username: "flaguser", token: "flagtoken" });
  });

  test("auth flag with colon in token", async () => {
    const creds = await resolveCredentials("https://ci.example.com", "user:token:with:colons");
    expect(creds).toEqual({ username: "user", token: "token:with:colons" });
  });

  test("invalid auth flag throws", async () => {
    expect(resolveCredentials("https://ci.example.com", "nocolon")).rejects.toThrow(
      "Invalid -auth format",
    );
  });

  test("env vars used when no auth flag", async () => {
    process.env.JENKINS_USER_ID = "envuser";
    process.env.JENKINS_API_TOKEN = "envtoken";
    const creds = await resolveCredentials("https://ci.example.com");
    expect(creds).toEqual({ username: "envuser", token: "envtoken" });
  });
});

File: internal/auth/auth_test.go
package auth_test

import (
	"os"
	"testing"

	"github.com/nwp/jenkins-cli/internal/auth"
)

func TestResolveCredentials_AuthFlag(t *testing.T) {
	creds, err := auth.ResolveCredentials(auth.ResolveOptions{
		AuthFlag:  "admin:mytoken",
		ServerURL: "http://jenkins.example.com",
	})
	if err != nil {
		t.Fatal(err)
	}
	if creds.Username != "admin" || creds.Token != "mytoken" {
		t.Errorf("got %+v", creds)
	}
}

func TestResolveCredentials_AuthFlag_ColonInToken(t *testing.T) {
	// Token may itself contain colons; only split on first colon.
	creds, err := auth.ResolveCredentials(auth.ResolveOptions{
		AuthFlag: "admin:tok:en:with:colons",
	})
	if err != nil {
		t.Fatal(err)
	}
	if creds.Username != "admin" || creds.Token != "tok:en:with:colons" {
		t.Errorf("got %+v", creds)
	}
}

func TestResolveCredentials_AuthFlag_Invalid(t *testing.T) {
	_, err := auth.ResolveCredentials(auth.ResolveOptions{AuthFlag: "nocolon"})
	if err == nil {
		t.Error("expected error for missing colon")
	}
}

func TestResolveCredentials_EnvVars(t *testing.T) {
	t.Setenv("JENKINS_USER_ID", "envuser")
	t.Setenv("JENKINS_API_TOKEN", "envtoken")

	creds, err := auth.ResolveCredentials(auth.ResolveOptions{ServerURL: "http://j.example.com"})
	if err != nil {
		t.Fatal(err)
	}
	if creds.Username != "envuser" || creds.Token != "envtoken" {
		t.Errorf("got %+v", creds)
	}
}

func TestResolveCredentials_AuthFlagTakesPriorityOverEnv(t *testing.T) {
	t.Setenv("JENKINS_USER_ID", "envuser")
	t.Setenv("JENKINS_API_TOKEN", "envtoken")

	creds, err := auth.ResolveCredentials(auth.ResolveOptions{
		AuthFlag: "flaguser:flagtoken",
	})
	if err != nil {
		t.Fatal(err)
	}
	if creds.Username != "flaguser" {
		t.Errorf("auth flag should take priority, got %+v", creds)
	}
}

func TestResolveCredentials_NoCredsNoTTY(t *testing.T) {
	// Make sure JENKINS_ env vars are unset
	os.Unsetenv("JENKINS_USER_ID")
	os.Unsetenv("JENKINS_API_TOKEN")

	// Override TTY detection so the TUI never launches in tests
	orig := auth.IsTTY
	auth.IsTTY = func() bool { return false }
	defer func() { auth.IsTTY = orig }()

	// No stored creds (empty server URL), no TTY → expect error
	_, err := auth.ResolveCredentials(auth.ResolveOptions{ServerURL: ""})
	if err == nil {
		t.Error("expected error when no credentials available and no TTY")
	}
}
